File: Army/Soldier.cpp
#include "Soldier.hpp"
#include "Attack.hpp"

Soldier::Soldier(const std::string& name): Unit(name) {}

Soldier::~Soldier() {}

void Soldier::takeDamage(int dmg) {
    this->getHealth().takeDamage(dmg);
}

void Soldier::attack(Unit& enemy) {
    this->getAttack().attack(enemy, *this);
}

void Soldier::counterAttack(Unit& enemy) {
    this->getAttack().counterAttack(enemy, *this);
}

void Soldier::transform(Unit& enemy) {
    this->getAbility().transform(enemy, *this);
}

File: Army(java)/Soldier.java
public class Soldier extends Unit {
    public Soldier(String name) {
        super(name);
    }

    public Soldier() {}

    @Override
    public void takeDamage(int dmg) {
        this.health.takeDamage(dmg);
    }

    @Override
    public void attack(Unit enemy) {
        this.attack.attack(enemy);
    }

    @Override
    public void counterAttack(Unit enemy) {
        this.attack.counterAttack(enemy);
    }

    @Override
    public void transform(Unit enemy) {
        this.ability.transform(enemy);
    }

    @Override
    public void transform() {
        this.ability.transform();
    }
}